datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnector.java

```diff
@@ -109,4 +141,143 @@ public void postDatastreamInitialize(Datastream datastream, List<Datastream> all
       throws DatastreamValidationException {
     _groupIdConstructor.populateDatastreamGroupIdInMetadata(datastream, allDatastreams, Optional.of(LOG));
   }
+
+  @Override
+  public void onPartitionChange(java.util.function.Consumer<DatastreamGroup> callback) {
+    _partitionChangeCallback = callback;
+  }
+
+  @Override
+  public void stop() {
+    super.stop();
+    _shutdown = true;
+    _partitionDiscoveryThreadMap.values().forEach(Thread::interrupt);
+  }
+
+  @Override
+  public Map<String, Optional<DatastreamPartitionsMetadata>> getDatastreamPartitions() {
+    Map<String, Optional<DatastreamPartitionsMetadata>> datastreams = new HashMap<>();
+    _partitionDiscoveryThreadMap.forEach((s, partitionDiscoveryThread) -> {
+      if (partitionDiscoveryThread._initialized) {
+        datastreams.put(s, Optional.of(new DatastreamPartitionsMetadata(s,
+            partitionDiscoveryThread._subscribedPartitions)));
+      } else {
+        datastreams.put(s, Optional.empty());
+      }
+    });
+    return datastreams;
+  }
+
+
+  @Override
+  public void handleDatastream(List<DatastreamGroup> datastreamGroups) {
+    List<String> dgNames = datastreamGroups.stream().map(DatastreamGroup::getTaskPrefix).collect(Collectors.toList());
+    List<String> obsoleteDgs = new ArrayList<>(_partitionDiscoveryThreadMap.keySet());
+    obsoleteDgs.removeAll(dgNames);
+    obsoleteDgs.forEach(name -> {
+      LOG.info("remove datastream group {}", name);
+      Optional.ofNullable(_partitionDiscoveryThreadMap.remove(name)).ifPresent(Thread::interrupt);
+    });
+
+    datastreamGroups.forEach(datastreamGroup -> {
+      String datastreamGroupName = datastreamGroup.getTaskPrefix();
+      PartitionDiscoveryThread partitionDiscoveryThread;
+      if (_partitionDiscoveryThreadMap.containsKey(datastreamGroupName)) {
+        partitionDiscoveryThread = _partitionDiscoveryThreadMap.get(datastreamGroupName);
+        partitionDiscoveryThread.setDatastreamGroup(datastreamGroup);
+      } else {
+        //We dont need to register the thread if no partition callback is available
+        if (_partitionChangeCallback != null) {
+          partitionDiscoveryThread =
+              new PartitionDiscoveryThread(datastreamGroup);
+          partitionDiscoveryThread.start();
+          _partitionDiscoveryThreadMap.put(datastreamGroupName, partitionDiscoveryThread);
+          LOG.info("DatastreamChangeListener for {} registered", datastreamGroupName);
+        }
+      }
+    });
+  }
+
+  class PartitionDiscoveryThread extends Thread {
+    private Consumer<?, ?> _consumer;
+    private DatastreamGroup _datastreamGroup;
+    private List<String> _subscribedPartitions = new ArrayList<>();
```

Comment:
Yes, subscribedPartitions is replaced with Collections.synchronizedList(), and the list itself is actually immutable. I'm going to replace the initialization part with Collections.EmptyList() to avoid confusion.